_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+```typescript
+export class ActivityApiComponent {
+
+    constructor(private store: Store<State>) {
+    }
+
+    getActivity(): void {
+        this.store.dispatch(getActivity());
+    }
+}
+```
+Jak widać w konstruktorze, wstrzykujemy obiekt Store, na którym wykonujemy metodę `dispatch` podając w jej argumencie typ akcji. Na tym kończy się odpowiedzialność komponentu.
+
+2. Wysłaną akcję przechwytuje efekt, który, zgodnie z kodem zaprezentowanym wcześniej, wysyła żądanie do API, a otrzymawszy odpowiedź przekazuje ją jako argument nowej akcji `Add Activity`.
+3. Akcja typu `Add Activity` jest przechwytywana przez reducer, który wyłuskuje z niej przekazaną z API odpowiedź i dodaje do store.
+4. W ostatniej kolejności do gry wchodzi selektor, który wykrywa zmianę store wynikającą z działania reducera. Użycie selektora w komponencie może wyglądać następująco:
+
+```typescript
+export class ActivityListComponent implements OnInit {
+
+  activities: ActivityItemModel[] = [];
+
+  constructor(private store: Store<State>) {
+  }
+
+  ngOnInit(): void {
+    this.store.select(getActivities)
+      .subscribe(activities => this.activities = activities);
+  }
+
+}
+```
+W metodzie `ngOnInit` wskazujemy, że chcemy nasłuchiwać na zmiany store przy pomocy selektora `getActivities`, a następnie przy każdej takiej zmianie aktualizujemy lokalnie utrzymywaną tablicę, którą można następnie wykorzystać w ciele komponentu.
+
+Tym samym cały proces dobiega końca, a kolejne kliknięcie przycisku wywoła go od nowa.
+
+## Inne podejścia
+Warto wspomnieć, że tak, jak dla Angulara istnieje biblioteka NgRx, tak również dla pozostałych frameworków z tzw. wielkiej trójcy znajdziemy implementacje architektury Redux:
+- Redux dla Reacta
+- Vuex dla Vue
+
+Należy również zaznaczyć, że Redux nie jest jedynym sposobem na zarządzanie stanem. Można tutaj wymienić takie alternatywy, jak:
+* [MobX](https://mobx.js.org/README.html) 
+* [Cycle.js](https://cycle.js.org/)
+* [Flux](https://github.com/facebook/flux) - warto wiedzieć, że Flux był protoplastą Reduxa
+* [Apollo](https://www.apollographql.com/)
+
+## Podsumowanie
+Łatwo zauważyć, że na pierwszy ogień NgRx, czy szerzej - Redux - potrafią nieco przytłoczyć ilością kodu, którą trzeba napisać, by nawet drobne funkcjonalności działały. Jest to jeden z największych zarzutów wobec tego rozwiązania, więc jeśli takie były Twoje odczucia podczas czytania tego artykułu - gratuluję krytycznego myślenia! Zauważmy jednak, że po przebrnięciu przez początkowe trudności zostajemy z aplikacją, którą bardzo łatwo możemy otestować, rozszerzać i której działanie jest jasno zdefiniowane.
+
+Mam nadzieję, że tym artykułem zainspirowałem nieco do zainteresowania się tematem zarządzania stanem aplikacji frontendowych. Serdecznie zapraszam do rozpoznawania tematu we własnym, gdyż przedstawiłem tu zaledwie wierzchołek możliwości, jakie zapewnia Redux i NgRx.
```

Comment:
Serdecznie zapraszam do rozpoznawania tematu we własnym... 
coś chyba zjadło ostatnie słowo